File: packages/ios/Sources/Components/FieldWrapper/FieldWrapper.swift
import SwiftUI

public struct KozmosFieldWrapper<Content: View>: View {
    public let label: String?
    public let description: String?
    public let optionalText: String?
    public let required: Bool
    public let hideLabel: Bool
    public let status: KozmosInputStatus
    public let error: Bool
    public let helperText: String?
    public let errorMessage: String?
    private let content: Content

    public init(
        label: String? = nil,
        description: String? = nil,
        optionalText: String? = nil,
        required: Bool = false,
        hideLabel: Bool = false,
        status: KozmosInputStatus = .default,
        error: Bool = false,
        helperText: String? = nil,
        errorMessage: String? = nil,
        @ViewBuilder content: () -> Content
    ) {
        self.label = label
        self.description = description
        self.optionalText = optionalText
        self.required = required
        self.hideLabel = hideLabel
        self.status = status
        self.error = error
        self.helperText = helperText
        self.errorMessage = errorMessage
        self.content = content()
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing75) {
            if shouldShowLabelRow {
                HStack(alignment: .firstTextBaseline, spacing: KozmosDimensions.primitivesLayoutSpacing150) {
                    HStack(alignment: .firstTextBaseline, spacing: 2) {
                        if let label = label {
                            Text(label)
                                .font(.subheadline.weight(.semibold))
                                .foregroundColor(labelColor)
                        }
                        if required {
                            Text("*")
                                .font(.subheadline.weight(.semibold))
                                .foregroundColor(KozmosColors.primitivesColorsEmotionalDanger600)
                        }
                    }

                    Spacer(minLength: KozmosDimensions.primitivesLayoutSpacing150)

                    if let optionalText = optionalDisplayText {
                        Text(optionalText)
                            .font(.caption)
                            .foregroundColor(metaColor)
                    }
                }
            }

            if let description = description, !description.isEmpty {
                Text(description)
                    .font(.subheadline)
                    .foregroundColor(metaColor)
                    .fixedSize(horizontal: false, vertical: true)
            }

            content
                .frame(maxWidth: .infinity, alignment: .leading)

            if let supportingText = supportingText, !supportingText.isEmpty {
                Text(supportingText)
                    .font(.subheadline)
                    .foregroundColor(supportingTextColor)
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private var effectiveStatus: KozmosInputStatus {
        error ? .error : status
    }

    private var supportingText: String? {
        errorMessage ?? helperText
    }

    private var shouldShowLabelRow: Bool {
        !hideLabel && (label != nil || optionalDisplayText != nil)
    }

    private var optionalDisplayText: String? {
        guard !required else { return nil }
        guard let optionalText = optionalText, !optionalText.isEmpty else { return nil }
        return optionalText
    }

    private var labelColor: Color {
        switch effectiveStatus {
        case .error:
            return KozmosColors.primitivesColorsEmotionalDanger600
        case .warning:
            return KozmosColors.primitivesColorsEmotionalAlert600
        case .success:
            return KozmosColors.primitivesColorsEmotionalSuccess600
        case .default:
            return KozmosColors.primitivesColorsForeground100
        }
    }

    private var supportingTextColor: Color {
        switch effectiveStatus {
        case .error:
            return KozmosColors.primitivesColorsEmotionalDanger600
        case .warning:
            return KozmosColors.primitivesColorsEmotionalAlert600
        case .success:
            return KozmosColors.primitivesColorsEmotionalSuccess600
        case .default:
            return KozmosColors.primitivesColorsForeground500
        }
    }

    private var metaColor: Color {
        KozmosColors.primitivesColorsForeground500
    }
}

public typealias KozmosFormField<Content: View> = KozmosFieldWrapper<Content>

File: packages/android/src/main/java/com/kozmos/components/FieldWrapper/FieldWrapper.kt
package com.kozmos.components.fieldwrapper

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.width
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import com.kozmos.components.input.KozmosInputStatus
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.tokens.KozmosThemeTokens

@Composable
fun KozmosFieldWrapper(
    modifier: Modifier = Modifier,
    label: String? = null,
    description: String? = null,
    optionalText: String? = null,
    required: Boolean = false,
    hideLabel: Boolean = false,
    status: KozmosInputStatus = KozmosInputStatus.Default,
    error: Boolean = false,
    helperText: String? = null,
    errorMessage: String? = null,
    content: @Composable () -> Unit
) {
    val effectiveStatus = if (error) KozmosInputStatus.Error else status
    val supportingText = errorMessage ?: helperText
    val statusColor = fieldStatusColor(effectiveStatus)
    val labelColor = statusColor ?: MaterialTheme.colorScheme.onSurface
    val metaColor = KozmosThemeTokens.primitivesColorsForeground500
    val optionalDisplayText = optionalText?.takeUnless { required || it.isBlank() }
    val showLabelRow = !hideLabel && (!label.isNullOrBlank() || optionalDisplayText != null)

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing75)
    ) {
        if (showLabelRow) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Row(
                    modifier = Modifier.weight(1f),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    if (!label.isNullOrBlank()) {
                        Text(
                            text = label,
                            style = MaterialTheme.typography.bodyMedium,
                            color = labelColor
                        )
                    }
                    if (required) {
                        Spacer(modifier = Modifier.width(KozmosDimensions.primitivesLayoutSpacing25))
                        Text(
                            text = "*",
                            style = MaterialTheme.typography.bodyMedium,
                            color = KozmosThemeTokens.primitivesColorsEmotionalDanger600
                        )
                    }
                }

                if (optionalDisplayText != null) {
                    Text(
                        text = optionalDisplayText,
                        style = MaterialTheme.typography.labelSmall,
                        color = metaColor
                    )
                }
            }
        }

        if (!description.isNullOrBlank()) {
            Text(
                text = description,
                style = MaterialTheme.typography.bodyMedium,
                color = metaColor
            )
        }

        content()

        if (!supportingText.isNullOrBlank()) {
            Text(
                text = supportingText,
                style = MaterialTheme.typography.bodyMedium,
                color = statusColor ?: metaColor
            )
        }
    }
}

@Composable
fun KozmosFormField(
    modifier: Modifier = Modifier,
    label: String? = null,
    description: String? = null,
    optionalText: String? = null,
    required: Boolean = false,
    status: KozmosInputStatus = KozmosInputStatus.Default,
    helperText: String? = null,
    content: @Composable () -> Unit
) {
    KozmosFieldWrapper(
        modifier = modifier,
        label = label,
        description = description,
        optionalText = optionalText,
        required = required,
        status = status,
        helperText = helperText,
        content = content
    )
}

@Composable
private fun fieldStatusColor(status: KozmosInputStatus): Color? {
    return when (status) {
        KozmosInputStatus.Error -> KozmosThemeTokens.primitivesColorsEmotionalDanger600
        KozmosInputStatus.Warning -> KozmosThemeTokens.primitivesColorsEmotionalAlert600
        KozmosInputStatus.Success -> KozmosThemeTokens.primitivesColorsEmotionalSuccess600
        KozmosInputStatus.Default -> null
    }
}
